File: 12주차/NH/Diameter of Binary Tree.js
function diameterOfBinaryTree(root) {
    let max=0;
    function self(root){
        if(root==null)return 0;
        
        let left=self(root.left);
        let right=self(root.right);
        
        if(left+right>max)max=left+right;
        max=Math.max(max,left+right);
        
        return Math.max(left,right)+1;
    }
    self(root);
    return max;
};

File: 12주차/YH/diameterOfBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        def depth(root) :
            if not root : return 0
            depthL = depth(root.left)
            depthR = depth(root.right)
            answer[0] = max(answer[0], depthL + depthR)
            
            return 1 + max(depthL, depthR)
        
        answer = [0]
        depth(root)
        return answer[0]